File: src/Java/pipeline-core/src/main/java/com/pipeline/core/PipelineError.java
package com.pipeline.core;

import java.util.Objects;

public record PipelineError(
    String pipelineName,
    StepPhase phase,
    int actionIndex,
    String actionName,
    Exception exception
) {
  public PipelineError {
    pipelineName = Objects.requireNonNull(pipelineName, "pipelineName");
    phase = Objects.requireNonNull(phase, "phase");
    if (actionIndex < 0) throw new IllegalArgumentException("actionIndex must be >= 0");
    actionName = Objects.requireNonNull(actionName, "actionName");
    exception = Objects.requireNonNull(exception, "exception");
  }
}

File: src/CSharp/pipeline_services/core/PipelineError.cs
using System;

namespace PipelineServices.Core;

public sealed class PipelineError
{
    public PipelineError(
        string pipelineName,
        StepPhase phase,
        int actionIndex,
        string actionName,
        Exception exception)
    {
        PipelineName = pipelineName ?? throw new ArgumentNullException(nameof(pipelineName));
        Phase = phase;
        if (actionIndex < 0)
        {
            throw new ArgumentOutOfRangeException(
                nameof(actionIndex),
                "actionIndex must be >= 0");
        }
        ActionIndex = actionIndex;
        ActionName = actionName ?? throw new ArgumentNullException(nameof(actionName));
        Exception = exception ?? throw new ArgumentNullException(nameof(exception));
    }

    public string PipelineName { get; }

    public StepPhase Phase { get; }

    public int ActionIndex { get; }

    public string ActionName { get; }

    public Exception Exception { get; }

    [Obsolete("Use ActionIndex.")]
    public int StepIndex => ActionIndex;

    [Obsolete("Use ActionName.")]
    public string StepName => ActionName;
}
